src/sss_client/common.c

```diff
@@ -684,21 +684,28 @@ static enum sss_status sss_cli_check_socket(int *errnop,
                                             int timeout)
 {
     static pid_t mypid;
-    struct stat mysb;
+    static struct stat selfsb;
+    struct stat mypid_sb, myself_sb;
     int mysd;
     int ret;
 
-    if (getpid() != mypid) {
-        ret = fstat(sss_cli_sd, &mysb);
+    ret = lstat("/proc/self/", &myself_sb);
+
+    if (getpid() != mypid || (ret == 0 && myself_sb.st_ino != selfsb.st_ino)) {
+        ret = fstat(sss_cli_sd, &mypid_sb);
         if (ret == 0) {
-            if (S_ISSOCK(mysb.st_mode) &&
-                mysb.st_dev == sss_cli_sb.st_dev &&
-                mysb.st_ino == sss_cli_sb.st_ino) {
+            if (S_ISSOCK(mypid_sb.st_mode) &&
+                mypid_sb.st_dev == sss_cli_sb.st_dev &&
+                mypid_sb.st_ino == sss_cli_sb.st_ino) {
                 sss_cli_close_socket();
             }
         }
         sss_cli_sd = -1;
         mypid = getpid();
+        ret = lstat("/proc/self/", &selfsb);
```

Comment:
Hi,
 My issue is: https://github.com/SSSD/sssd/issues/6592
I do not know the use of `mypid`, but `mypid` causes this issue.  so I imitate the writing method of `mypid` and add the `selfsb` and` myself_sb` to reduce the probability of problems caused by pid loopback.

Reply:
Hi,

yes, I understand this. However, currently `getpid()` and with your patch `lstat()` as well are called twice to get the same information. Since those are system-call both need some time to execute and since the second call is redundant it would help to avoid it.

The `getpid()` related code it quite old and at the time it was written it might have even been cheap because glibc had a pid cache which helped to avoid a system-call. But this changed are there is currently also no alternatives, e.g.  as a vDSO call, available, see e.g.  https://bugzilla.redhat.com/show_bug.cgi?id=1469757 and the links in the ticket for details. I wasn't aware of this until reviewing your patch as well and thought that since your patch touched this area of the code it can be improved as part  of your patch.

If @alexey-tikhonov, agrees we can keep your patch as it is but open a new ticket to remove the second `getpid() and `lstat()` call in a separate PR. But feel free to do the change yourself.

bye,
Sumit

Reply:
Nice catch. I think at least 2nd `lstat()` can be avoided within this patch.

Also I think we should really consider per-transaction sockets, as you proposed in other ticket...